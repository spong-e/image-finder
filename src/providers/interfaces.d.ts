interface DetailsState {
  details: Details;
}

interface Details {
  firstName: string;
  surname: string;
  topic: string;
  thumbnail: string;
}

interface DetailsAction {
  type: string;
  payload: Details;
}
