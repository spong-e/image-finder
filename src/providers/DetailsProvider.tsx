import * as React from "react";
import { createContext, Dispatch, ReactNode, useReducer } from "react";
import detailsReducer from "./DetailsReducer";

const DetailsContext = createContext<DetailsState | undefined>(undefined);

const DetailsDispatchContext = createContext<
  Dispatch<DetailsAction> | undefined
>(undefined);

interface DetailsProviderProps {
  children: ReactNode;
}

function DetailsProvider({ children }: DetailsProviderProps) {
  const initialState: DetailsState = {
    details: {
      firstName: "",
      surname: "",
      topic: "",
      thumbnail: "",
    },
  };

  const [state, dispatch] = useReducer(detailsReducer, initialState);

  // NOTE: you *might* need to memoize this value
  // Learn more in http://kcd.im/optimize-context

  const value = { state, dispatch };
  return (
    <DetailsContext.Provider value={state}>
      <DetailsDispatchContext.Provider value={dispatch}>
        <>
          {JSON.stringify(state, null, 2)}
          {children}
        </>
      </DetailsDispatchContext.Provider>
    </DetailsContext.Provider>
  );
}

export { DetailsProvider, DetailsContext, DetailsDispatchContext };
