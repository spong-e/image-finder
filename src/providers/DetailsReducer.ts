import produce from "immer";
import { Reducer } from "react";
import axios from "axios";
import { createApi } from "unsplash-js";
import { ACTION_TYPES } from "./constants";

const detailsReducer: Reducer<DetailsState, DetailsAction> = produce(
  (draft: DetailsState, action: DetailsAction) => {
    const { type, payload } = action;
    switch (type) {
      case ACTION_TYPES.SET_DETAILS: {
        draft.details = payload;

        return;
      }

      default: {
        throw new Error(`Unhandled action type: ${action.type}`);
      }
    }
  }
);

export default detailsReducer;
