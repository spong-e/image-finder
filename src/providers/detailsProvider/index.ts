export * from "./constants";
export * from "./context";
export * from "./hooks";
