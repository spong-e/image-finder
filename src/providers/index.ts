export { DetailsProvider } from "./DetailsProvider";
export { default as useDetails } from "./useDetails";
export { default as useDetailsActions } from "./useDetailsActions";
