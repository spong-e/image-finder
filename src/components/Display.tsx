import { FunctionComponent } from "react";
import { Link } from "react-router-dom";
import { Card, Image, Message } from "semantic-ui-react";

import { ROUTES } from "../constants";
import { useDetails } from "../providers";

const Display: FunctionComponent = () => {
  const details = useDetails();

  if (
    !details ||
    !details.firstName ||
    !details.lastName ||
    !details.thumbnail
  ) {
    return (
      <>
        <Message negative>
          <Message.Header>
            Something has gone wrong with your search
          </Message.Header>
          <Link to={ROUTES.ENTRY}>Start again</Link>
        </Message>
      </>
    );
  }

  const { thumbnail, firstName, lastName } = details;

  return (
    <>
      <Card>
        <Card.Content>
          <Image
            src={thumbnail}
            floated="right"
            size="small"
            data-testid="cardThumbnail"
          />

          <Card.Header data-testid="cardHeader">
            {firstName} {lastName}
          </Card.Header>
        </Card.Content>
      </Card>

      <Link to={ROUTES.ENTRY}>Start again</Link>
    </>
  );
};

export default Display;
