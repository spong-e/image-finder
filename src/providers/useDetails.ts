import { FunctionComponent, useContext } from "react";
import { DetailsContext } from "./DetailsProvider";

const useDetails = () => {
  const detailsContext = useContext(DetailsContext);
  if (!detailsContext)
    throw new Error(
      "No DetailsContext.Provider found when calling useGridItemContext."
    );

  const { details } = detailsContext;

  return details;
};

export default useDetails;
