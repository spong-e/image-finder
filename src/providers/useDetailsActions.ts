import { useContext } from "react";
import { ACTION_TYPES } from "./constants";
import { DetailsDispatchContext } from "./DetailsProvider";

const useDetailsActions = () => {
  const detailsDispatchContext = useContext(DetailsDispatchContext);
  if (!detailsDispatchContext)
    throw new Error(
      "No DetailsDispatchContext.Provider found when calling useGridItemContext."
    );

  const set = (details: Details) => {
    detailsDispatchContext({
      type: ACTION_TYPES.SET_DETAILS,
      payload: details,
    });
  };

  return { set };
};

export default useDetailsActions;
