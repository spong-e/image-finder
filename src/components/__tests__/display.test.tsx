import { MemoryRouter } from "react-router-dom";

import { render, screen } from "@testing-library/react";

import { useDetails } from "../../providers";
import Display from "../Display";

import "@testing-library/jest-dom";

const details: Details = {
  firstName: "Test",
  lastName: "User",
  thumbnail: "http://tempuri/thumbnail",
  topic: "testing",
};

jest.mock("../../providers");

describe("Display component", () => {
  describe("Details returned", () => {
    beforeEach(async () => {
      (useDetails as jest.Mock).mockReturnValue(details);
    });

    it("Given useDetails returns data, component renders", async () => {
      const { container } = render(
        <MemoryRouter>
          <Display />
        </MemoryRouter>
      );

      const thumbnail = await screen.findByTestId("cardThumbnail");
      const header = await screen.findByTestId("cardHeader");

      expect(thumbnail).toHaveAttribute("src", details.thumbnail);
      expect(header).toHaveTextContent(
        `${details.firstName} ${details.lastName}`
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe("No details returned", () => {
    beforeEach(async () => {
      (useDetails as jest.Mock).mockReturnValue(null);
    });

    it("Given useDetails returns null, component does not render", async () => {
      const { container } = render(
        <MemoryRouter>
          <Display />
        </MemoryRouter>
      );

      expect(container).toMatchSnapshot();
    });
  });
});
