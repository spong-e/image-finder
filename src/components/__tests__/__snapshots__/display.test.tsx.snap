// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display component Details returned Given useDetails returns data, component renders 1`] = `
<div>
  <div
    class="ui card"
  >
    <div
      class="content"
    >
      <img
        class="ui small right floated image"
        data-testid="cardThumbnail"
        src="http://tempuri/thumbnail"
      />
      <div
        class="header"
        data-testid="cardHeader"
      >
        Test
         
        User
      </div>
    </div>
  </div>
  <a
    href="/"
  >
    Start again
  </a>
</div>
`;

exports[`Display component No details returned Given useDetails returns null, component does not render 1`] = `
<div>
  <div
    class="ui negative message"
  >
    <div
      class="header"
    >
      Something has gone wrong with your search
    </div>
    <a
      href="/"
    >
      Start again
    </a>
  </div>
</div>
`;
